seqr/utils/middleware.py

```diff
@@ -56,6 +61,70 @@ def __init__(self, errors, warnings=None):
 
 ERROR_LOG_EXCEPTIONS = {InvalidIndexException}
 
+
+class CheckProgrammaticAccessMiddleware(MiddlewareMixin):
+
+    def process_request(self, request):
+        func, _, _ = resolve(request.path)
+        request.programmatic_access = isinstance(func, ProgrammaticAccess)
+
+
+class DisableCSRFProgrammaticAccessMiddleware(MiddlewareMixin):
+
+    def process_view(self, request, callback, callback_args, callback_kwargs):
+        assert hasattr(request, 'programmatic_access'), (
+            'The seqr DisableCSRFProgrammaticAccess middleware requires '
+            'CheckProgrammaticAccessMiddleware middleware to be installed. '
+            'Edit your MIDDLEWARE setting to insert '
+            '"seqr.utils.middleware.CheckProgrammaticAccessMiddleware" before '
+            '"seqr.utils.middleware.DisableCSRFProgrammaticAccessMiddleware".'
+        )
+
+        if request.programmatic_access:
```

Comment:
Should the exemption be even more restricted to safe methods only to match logic in https://docs.djangoproject.com/en/3.2/_modules/django/middleware/csrf/?

Also, should remove comment about `request._dont_enforce_csrf_checks`, it's not correct anyway and should only be used under testing contexts.

Reply:
If I understand correctly, CSRF doesn't apply on safe methods right? We'd want to disable CSRF because we'll want to change state in the future. Maybe we could have ReadOnlyServiceAccountAccess and ReadWriteServiceAccountAccess - but we're already restricting these ServiceAccounts to only routes specifically classed as ServiceAccount (ie: they cannot authenticate and CSRF would block them for non-service-account views.


>  Request methods are considered "safe" if their defined semantics are essentially read-only
>      _Source: [RFC7231#section-4.2.2](https://datatracker.ietf.org/doc/html/rfc7231#:~:text=%20request%20methods%20are%20considered%20%22safe%22%20if%20their%20defined%20semantics%20are%20essentially%20read-only)_

```python
# Assume that anything not defined as 'safe' by RFC7231 needs protection`
if request.method not in ('GET', 'HEAD', 'OPTIONS', 'TRACE'):

# Source: https://docs.djangoproject.com/en/3.2/_modules/django/middleware/csrf/
```

Reply:
I guess what I'm saying is rather than disable CSRF for all HTTP methods, just disable if it's not a safe method.  This would make it consistent with the logic in `CsrfViewMiddleware` which subsequently gets called.